extend PythonGapicClientPipeline to optionally publish to devpi

mvashishtha opened the pull request:
PythonGapicClientPipeline now has devpi publishing tasks, so that its output can be published to devpi, with username, server URL, and index provided in a config yaml (though these tasks are only added to the flow if the index is not "none"). I also edited CleanUpTask to extend task_base rather than task.

garrettjonesgoogle commented:
Could you clean up your commit message?

ethanbao commented on test/test_pipeline_baseline.py:
```diff
@@ -73,6 +73,14 @@ def _test_baseline(task_name, test_name, language, output_dir, mock_check_call,
         'ignore_base': False,
         'final_repo_dir': output_dir}
+    publish_kwargs = {
+        'publish_server_url': 'https://gapi-pypi.appspot.com/',
```
Let's use some fake data here.

mvashishtha replied:
used fake data

ethanbao commented on pipeline/pipelines/code_generation_pipeline.py:
```diff
@@ -16,7 +16,8 @@
 from pipeline.pipelines import pipeline_base
-from pipeline.tasks import protoc_tasks, gapic_tasks, format_tasks
+from pipeline.tasks import (protoc_tasks, gapic_tasks, format_tasks,
+                            publish_python_gapic_tasks)
```
I suggest we rename the name from publish_python_gapi_tasks to publish_tasks.

mvashishtha replied:
done.

geigerj commented on pipeline/pipelines/code_generation_pipeline.py:
```diff
@@ -103,10 +116,16 @@ def do_build_flow(self, **kwargs):
                  format_tasks.PythonFormatTask('PythonFormat', inject=kwargs),
                  gapic_tasks.GapicMergeTask('GapicMerge', inject=kwargs),
                  gapic_tasks.GapicPackmanTask('GapicPackman', inject=kwargs))
+        if kwargs['publish_env'] != "none":
```
Prefer single quotes to double quotes

geigerj replied:
I would also prefer that publishing was based on the existence/absence of the `publish_env` flag, rather than needing to set the flag to "none" to disable.

jmuk replied:
I also prefer not changing the construction of tasks by a kwargs.

Those tasks can take the 'publish_env' argument and do nothing if not.

mvashishtha replied:
changed to single quotes.  also changed to desired behavior (based on presence of publish_env) as mentioned by jmuk

geigerj commented on pipeline/pipelines/code_generation_pipeline.py:
```diff
@@ -103,10 +116,16 @@ def do_build_flow(self, **kwargs):
 
     def validate_kwargs(self, **kwargs):
-        _validate_codegen_kwargs(_VGEN_REQUIRED, **kwargs)
+        _validate_codegen_kwargs(_VGEN_REQUIRED+_PYTHON_PUB_REQUIRED, **kwargs)
```
There should be a space around the `+` operator. The linters that we've configured with `tox` -- surprised the linters we have configured with `tox` don't catch this...

geigerj commented on pipeline/tasks/publish_python_gapic_tasks.py:
```diff
@@ -0,0 +1,46 @@
+
+
+class DevpiIndexTask(task_base.TaskBase):
+    """ sets the devpi index for publishing: publish_env in
```
Remove leading space in the docstring, capitalize the first letter

mvashishtha replied:
done

geigerj replied:
I'm not seeing this change?

geigerj commented on pipeline/tasks/publish_python_gapic_tasks.py:
```diff
@@ -0,0 +1,46 @@
+
+class DevpiIndexTask(task_base.TaskBase):
+    """ sets the devpi index for publishing: publish_env in
+    config file must be dev, test, staging, or prod"""
```
Prefer to reference the variable `_PYTHON_PUB_ENVS` here rather than listing its values to reduce the risk of a stale docstring if we update the variable.

mvashishtha replied:
made reference to variable in pipeline/pipelines/code_generation_pipeline.py

jmuk commented on pipeline/tasks/publish_python_gapic_tasks.py:
```diff
@@ -0,0 +1,46 @@
+    """ sets the devpi index for publishing: publish_env in
+    config file must be dev, test, staging, or prod"""
+
+    def execute(self):
```
def execute(self, publish_server_url, publish_uname, publish_env):

Then, the taskflow framework can take care of arguments. You don't have to access to inject directly.

ethanbao replied:
+1

mvashishtha replied:
done

jmuk commented on pipeline/tasks/publish_python_gapic_tasks.py:
```diff
@@ -0,0 +1,46 @@
+
+
+class DevpiUploadTask(task_base.TaskBase):
+
```
add the docstring.

mvashishtha replied:
these tasks have been combined

jmuk commented on pipeline/tasks/publish_python_gapic_tasks.py:
```diff
@@ -0,0 +1,46 @@
+
+class DevpiUploadTask(task_base.TaskBase):
+
+    def execute(self):
```
Similarly, def execute(self, publish_output_folder)

jmuk commented on test/testdata/python_gapic_client_pipeline.baseline:
```diff
@@ -4,3 +4,5 @@ rm -rf {OUTPUT}/library-v1-gapic-gen-python
 test/fake-repos/toolkit/gradlew -p test/fake-repos/toolkit runSynchronizer -Pclargs=--source_path={OUTPUT},--generated_path={OUTPUT}/library-v1-gapic-gen-python,--baseline_path={OUTPUT}/baseline,--auto_merge,--auto_resolve
 gen-api-package --api_name=library/v1 -l python --gax_dir={OUTPUT} --template_root=templates/gax
+devpi use https://gapi-pypi.appspot.com/mvashishtha/dev
+devpi upload --no-vcs --from-dir {OUTPUT}
```
Add the newline at the end.
Also, after my commit of #23, this would be --from-dir {OUTPUT}/final

jmuk commented on pipeline/tasks/cleanup_tasks.py:
```diff
@@ -16,10 +16,10 @@
 
 
-class CleanUpTask(task.Task):
+class CleanUpTask(task_base.TaskBase):
```
cool fix! 👍

jmuk commented:
This will require a change on googleapis/googleapis repository (otherwise normal execute_pipeline will fail). Do you have that one?

Also -- it's not very clear how to override those values -- updating USAGE.rst file as well?

ethanbao commented on pipeline/tasks/publish_python_gapic_tasks.py:
```diff
@@ -0,0 +1,46 @@
+from pipeline.tasks import task_base
+
+
+class DevpiIndexTask(task_base.TaskBase):
```
Can we combine these two tasks into one?

mvashishtha replied:
combined.

mvashishtha commented:
@jmuk could you explain why this will require a change in the googleapis/googleapis repository? Why will execute_pipeline fail otherwise?

jmuk commented:
pipeline arguments come from the config files in googleapis/googleapis. If you add new items to kwargs and it's required at validate_kwargs(), those new arguments have to exist.

Please review USAGE.rst and make sure the sample commands work as-is.

geigerj commented:
It's a little confusing that other people's (merged) PRs are showing up in yours. It might be less confusing if you rebase all of your commits on top of HEAD of googleapis/artman

mvashishtha commented:
@jmuk @geigerj @ethanbao TALP. Responded to comments on initial commit and incorporated Jun's and Garrett's changes. Also added publishing to grpc pipeline instead of gapic pipeline, as ethan mentioned.

ethanbao commented on pipeline/pipelines/code_generation_pipeline.py:
```diff
@@ -30,6 +30,10 @@
+_PYTHON_PUB_REQUIRED = ['publish_server_url',
+                        'publish_uname',
+                        'publish_env']
+_PYTHON_PUB_ENVS = ['dev', 'test', 'staging', 'prod', 'none']
```
nit: We don't need 'none'.

mvashishtha replied:
'none' removed

jmuk commented on pipeline/tasks/publish_tasks.py:
```diff
@@ -0,0 +1,38 @@
+class DevpiUploadTask(task_base.TaskBase):
+    """Publishes to a devpi index: publish_env in config file must be
+    in _PYTHON_PUB_ENVS list in
+    pipeline/pipelines/code_generation_pipeline.py"""
```
nit: I don't think the latter part of this doccomment is necessary -- simply 'Publishes to a devpi index'.

mvashishtha replied:
done

geigerj commented on pipeline/pipelines/code_generation_pipeline.py:
```diff
@@ -56,6 +60,13 @@ def _validate_codegen_kwargs(extra_args, **kwargs):
 
+def _validate_publish_kwargs(**kwargs):
+    if kwargs['publish_env'] not in _PYTHON_PUB_ENVS:
+        raise ValueError("Invalid devpi publishing environment " +
```
- use single quotes
- prefer string format to concatentation; then you don't have to manually stringify.

E.g.,
'Invalid environment "{}"; choose from one of {}'.format(kwargs['publish_env'], _PYTHON_PUB_ENVS)

mvashishtha replied:
made both changes

geigerj commented on pipeline/tasks/publish_tasks.py:
```diff
@@ -0,0 +1,38 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+""" Tasks for publishing artman output """
```
Trim leading and trailing spaces

mvashishtha replied:
done

geigerj commented on test/test_pipeline_baseline.py:
```diff
@@ -51,8 +51,8 @@ def check_calls_match(expected_calls, actual_calls):
 @mock.patch('os.chdir')
-def _test_baseline(task_name, test_name, language, output_dir, mock_chdir,
-                   mock_check_call, mock_call, mock_gradle_task):
+def _test_baseline(task_name, test_name, language, output_dir, extra_kwargs,
```
Let's make extra_kwargs an optional keyword argument

mvashishtha replied:
That makes sense. I've been trying to figure out how to combine this change with the mock patches on this function, though. If extra_kwargs is an optional argument, it has to be the last parameter of the function definition, so we might have a header like `def _test_baseline(task_name, test_name, language, output_dir, mock_chdir, mock_check_call, mock_call, mock_gradle_task, extra_kwargs={})`. But the patches seem to patch starting from the last argument by default, so they try to patch extra_kwargs. How should I add the optional extra_kwargs argument so that the mocks still work?

geigerj replied:
Eeek. OK, maybe we should just leave this.

geigerj commented on test/test_pipeline_baseline.py:
```diff
@@ -106,24 +107,50 @@ def _test_python_baseline(task_name, test_name, tmpdir):
+    output_dir = str(tmpdir)
+
+    # Create an empty 'fake_output_api.py' in the output_dir. Do not invoke
+    #  'touch' command with subprocess.call() because it's mocked.
```
remove extra space

mvashishtha replied:
do you mean the empty line #119 in the new commit?

geigerj replied:
I meant that the second line in this comment is indented one space relative to the first line -- there are two spaces between the `#` and `'touch'`.

mvashishtha replied:
done

geigerj commented on test/test_pipeline_baseline.py:
```diff
@@ -106,24 +107,50 @@ def _test_python_baseline(task_name, test_name, tmpdir):
+    # Create an empty 'fake_output_api.py' in the output_dir. Do not invoke
+    #  'touch' command with subprocess.call() because it's mocked.
+    final_output_dir = os.path.join(output_dir, 'library-v1-gapic-gen-python')
+    if not os.path.exists(final_output_dir):
```
Why do we need to manipulate the filesystem? This should all be mocked.

mvashishtha replied:
the os code comes from the method _test_python_baseline in the current master branch (it was written by jun according to blame). I created _test_python_pub_baseline and _test_python_nopub_baseline based on _test_python_baseline, so I used this code. Should I figure out how to avoid manpiulating the filesystem?

geigerj replied:
Ah, I see where it comes from now. In that case, these methods should be refactored to remove the code duplication.

mvashishtha replied:
refactored

mvashishtha commented:
@geigerj @ethanbao @jmuk TALP

geigerj commented on pipeline/tasks/publish_tasks.py:
```diff
@@ -0,0 +1,42 @@
+            ['devpi',
+             'login',
+             '--password',
+             pypi_pwd,
```
I think this command is specific to devpi, rather than PyPI, so let's rename these variables `devpi_{server_url,pwd,uname}`

geigerj replied:
Never mind -- I see this was originally changed from devpi so that the input kwargs to the pipeline could be more general.

ethanbao replied:
I actually suggest everything be called pypi (by assuming devpi is just one implementation of PyPi). Therefore, the task name should be called PypiUploadTask. By doing that, we hide the implementation details from users.

geigerj replied:
That is true. The implementation will be different between "real" PyPI and Devpi. (AFAIK, real PyPI has no client like devpi does.) But to the user, it is best not to make that distinction.

geigerj commented:
LGTM

jmuk commented on pipeline/tasks/publish_tasks.py:
```diff
@@ -0,0 +1,42 @@
+        publish_url = pypi_server_url + pypi_uname + '/' + publish_env
+        subprocess.check_call(
+            ['devpi',
+             'login',
```
I am not sure why login should be invoked every time.
Can it be optional?  It's natural to use existing auth on the local environment.

ethanbao replied:
This might be a limitation of devpi, which doesn't allow passing username/password (or auth token) during publishing. That makes the implementation looks like logging in every time the task gets running.

On the other hand, if we don't do the login in this task, it would mean that whoever runs the pipeline for pypi publishing need to do `devpi login` manually. I feel that is not ideal either.

jmuk replied:
I see. Thinking about the remote execution environment, probably this is the right way to do (as for local execution, I think it might be great to reuse existing user's authentication though).

ethanbao replied:
Good point. For local execution, we might allow skip login if username/password is not passed. And do the validation before execution to make sure that existing user's auth is available.

jmuk commented:
LGTM -- probably you should address @ethanbao 's comment about devpi vs pypi naming, but overall it looks good.

mvashishtha commented:
@jmuk @geigerj @ethanbao 
Renamed DevpiUploadTask to PypiUpload task. PTAL

ethanbao commented:
LGTM. Go ahead and merge.
